File: backend-net10/Services/ClienteService.cs
using Microsoft.EntityFrameworkCore;
using backend_net10.Data;
using backend_net10.Models;

namespace backend_net10.Services;

public class ClienteService
{
    private readonly AppDbContext _context;

    public ClienteService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<List<Cliente>> ListAsync()
    {
        return await _context.Clientes.ToListAsync();
    }

    public async Task<Cliente> GetByIdAsync(string id)
    {
        var client = await _context.Clientes
            .Include(c => c.Propostas!)
                .ThenInclude(p => p.Empresa)
            .FirstOrDefaultAsync(c => c.Id == id);

        if (client == null)
        {
            throw new AppNotFoundException("cliente não encontrado");
        }

        return client;
    }

    public async Task<Cliente> CreateAsync(Cliente client)
    {
        if (string.IsNullOrWhiteSpace(client.Nome))
        {
            throw new AppBadRequestException("nome do cliente é obrigatório");
        }

        if (string.IsNullOrWhiteSpace(client.Email) || !client.Email.Contains('@'))
        {
            throw new AppBadRequestException("email do cliente é obrigatório ou inválido");
        }

        client.CPFCNPJ = Validation.CleanCPFCNPJ(client.CPFCNPJ);
        if (!Validation.IsValidCPFCNPJ(client.CPFCNPJ))
        {
            throw new AppBadRequestException("CPF ou CNPJ inválido");
        }

        var existing = await _context.Clientes.FirstOrDefaultAsync(c => c.CPFCNPJ == client.CPFCNPJ);
        if (existing != null)
        {
            throw new AppBadRequestException("cliente com este CPF/CNPJ já cadastrado");
        }

        client.Id = Guid.NewGuid().ToString();
        client.CreatedAt = DateTime.Now;
        client.UpdatedAt = DateTime.Now;

        _context.Clientes.Add(client);
        await _context.SaveChangesAsync();

        return client;
    }

    public async Task<Cliente> UpdateAsync(string id, Cliente input)
    {
        var client = await _context.Clientes.FirstOrDefaultAsync(c => c.Id == id);
        if (client == null)
        {
            throw new AppNotFoundException("cliente não encontrado");
        }

        if (string.IsNullOrWhiteSpace(input.Nome))
        {
            throw new AppBadRequestException("nome do cliente é obrigatório");
        }

        if (string.IsNullOrWhiteSpace(input.Email) || !input.Email.Contains('@'))
        {
            throw new AppBadRequestException("email do cliente é obrigatório ou inválido");
        }

        var cleanCPF = Validation.CleanCPFCNPJ(input.CPFCNPJ);
        if (cleanCPF != client.CPFCNPJ)
        {
            if (!Validation.IsValidCPFCNPJ(cleanCPF))
            {
                throw new AppBadRequestException("CPF ou CNPJ inválido");
            }

            var existing = await _context.Clientes.FirstOrDefaultAsync(c => c.CPFCNPJ == cleanCPF);
            if (existing != null && existing.Id != client.Id)
            {
                throw new AppBadRequestException("cliente com este CPF/CNPJ já cadastrado");
            }
            client.CPFCNPJ = cleanCPF;
        }

        client.Nome = input.Nome;
        client.Email = input.Email;
        client.Telefone = input.Telefone;
        client.Endereco = input.Endereco;
        client.UpdatedAt = DateTime.Now;

        await _context.SaveChangesAsync();
        return client;
    }

    public async Task DeleteAsync(string id)
    {
        var client = await _context.Clientes.FirstOrDefaultAsync(c => c.Id == id);
        if (client == null)
        {
            throw new AppNotFoundException("cliente não encontrado");
        }

        _context.Clientes.Remove(client);
        await _context.SaveChangesAsync();
    }
}

File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/ClienteService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.model.Cliente;
import br.com.portalproposta.backend.model.Validation;
import br.com.portalproposta.backend.repository.ClienteRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class ClienteService {

    private final ClienteRepository clienteRepository;

    @Autowired
    public ClienteService(ClienteRepository clienteRepository) {
        this.clienteRepository = clienteRepository;
    }

    public List<Cliente> list() {
        return clienteRepository.findAll();
    }

    public Cliente getById(String id) {
        return clienteRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("cliente não encontrado"));
    }

    @Transactional
    public Cliente create(Cliente client) {
        if (client.getNome() == null || client.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do cliente é obrigatório");
        }

        if (client.getEmail() == null || client.getEmail().trim().isEmpty() || !client.getEmail().contains("@")) {
            throw new AppBadRequestException("email do cliente é obrigatório ou inválido");
        }

        client.setCpfCnpj(Validation.cleanCPFCNPJ(client.getCpfCnpj()));
        if (!Validation.isValidCPFCNPJ(client.getCpfCnpj())) {
            throw new AppBadRequestException("CPF ou CNPJ inválido");
        }

        clienteRepository.findByCpfCnpj(client.getCpfCnpj()).ifPresent(existing -> {
            throw new AppBadRequestException("cliente com este CPF/CNPJ já cadastrado");
        });

        client.setId(UUID.randomUUID().toString());
        client.setCreatedAt(LocalDateTime.now());
        client.setUpdatedAt(LocalDateTime.now());

        return clienteRepository.save(client);
    }

    @Transactional
    public Cliente update(String id, Cliente input) {
        Cliente client = getById(id);

        if (input.getNome() == null || input.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do cliente é obrigatório");
        }

        if (input.getEmail() == null || input.getEmail().trim().isEmpty() || !input.getEmail().contains("@")) {
            throw new AppBadRequestException("email do cliente é obrigatório ou inválido");
        }

        String cleanCPF = Validation.cleanCPFCNPJ(input.getCpfCnpj());
        if (!cleanCPF.equals(client.getCpfCnpj())) {
            if (!Validation.isValidCPFCNPJ(cleanCPF)) {
                throw new AppBadRequestException("CPF ou CNPJ inválido");
            }

            clienteRepository.findByCpfCnpj(cleanCPF).ifPresent(existing -> {
                if (!existing.getId().equals(client.getId())) {
                    throw new AppBadRequestException("cliente com este CPF/CNPJ já cadastrado");
                }
            });
            client.setCpfCnpj(cleanCPF);
        }

        client.setNome(input.getNome());
        client.setEmail(input.getEmail());
        client.setTelefone(input.getTelefone());
        client.setEndereco(input.getEndereco());
        client.setUpdatedAt(LocalDateTime.now());

        return clienteRepository.save(client);
    }

    @Transactional
    public void delete(String id) {
        Cliente client = getById(id);
        clienteRepository.delete(client);
    }
}
